ethcore/engines/authority-round/src/randomness.rs

```diff
@@ -0,0 +1,245 @@
+/// |              |                              |               |
+/// |  Committed   +------------------------------>    Reveal     |
+/// |              |  enter reveal phase          |               |
+/// +--------------+                              +---------------+
+/// ```
+///
+///
+/// Phase transitions are performed by the smart contract and simply queried by the engine.
+///
+/// A typical case of using `RandomnessPhase` is:
+///
+/// 1. `RandomnessPhase::load()` the phase from the blockchain data.
+/// 2. Call `RandomnessPhase::advance()`.
+#[derive(Debug)]
+pub enum RandomnessPhase {
+	// NOTE: Some states include information already gathered during `load` (e.g. `our_address`,
+	//       `round`) for efficiency reasons.
+	/// Waiting for the next phase.
+	///
+	/// This state indicates either the successful revelation in this round or having missed the
+	/// window to make a commitment.
+	Waiting,
+	/// Indicates a commitment is possible, but still missing.
+	BeforeCommit { our_address: Address, round: U256 },
+	/// Indicates a successful commitment, waiting for the commit phase to end.
+	Committed,
+	/// Indicates revealing is expected as the next step.
+	Reveal { our_address: Address, round: U256 },
+}
+
+/// Phase loading error for randomness generation state machine.
+///
+/// This error usually indicates a bug in either the smart contract, the phase loading function or
+/// some state being lost.
+///
+/// The `LostSecret` and `StaleSecret` will usually result in punishment by the contract or the
+/// other validators.
+#[derive(Debug, Display)]
+pub enum PhaseError {
+	/// The smart contract reported a phase as both commitment and reveal phase.
+	#[display(fmt = "Inconsistent randomness phase information")]
+	PhaseConflict,
+	/// The smart contract reported that we already revealed something while still being in the
+	/// commit phase.
+	#[display(fmt = "Revealed during commit phase")]
+	RevealedInCommit,
+	/// Failed to load contract information.
+	#[display(fmt = "Error loading randomness contract information: {:?}", _0)]
+	LoadFailed(CallError),
+	/// Failed to load the stored secret.
+	#[display(fmt = "Failed to load secret from the randomness contract")]
+	StaleSecret,
+	/// Failed to encrypt stored secret.
+	#[display(fmt = "Failed to encrypt stored randomness secret: {}", _0)]
+	Crypto(CryptoError),
+	/// Failed to decrypt stored secret.
+	#[display(fmt = "Failed to decrypt stored randomness secret: {}", _0)]
+	Decrypt(CryptoError),
+	/// Failed to get the engine signer's public key.
+	#[display(fmt = "Failed to get the engine signer's public key")]
+	MissingPublicKey,
+}
+
+impl RandomnessPhase {
+	/// Determine randomness generation state from the contract.
+	///
+	/// Calls various constant contract functions to determine the precise state that needs to be
+	/// handled (that is, the phase and whether or not the current validator still needs to send
+	/// commitments or reveal secrets).
+	pub fn load(
```

Comment:
`current_collect_round()` is needed for `is_committed()` and `sent_reveal()` since they accept `_collectRound` parameter.

We use `is_committed()` and `sent_reveal()` to prevent duplicated calls to `commit_hash()` and `reveal_secret()` respectively.

To simplify that a bit we could utilize `is_commit_phase()` instead of `is_reveal_phase()` as `is_commit_phase() == !is_reveal_phase()`.

Reply:
@afck Let's use `is_commit_phase()` instead of `is_reveal_phase()`: `let is_reveal_phase = !is_commit_phase` in https://github.com/poanetwork/parity-ethereum/blob/727fce39a1b8c922db27533b9d38bc879f707a32/ethcore/engines/authority-round/src/randomness.rs#L146-L151

Reply:
`is_reveal_phase` was removed from the [`load` function](https://github.com/poanetwork/parity-ethereum/blob/4799aeb3d11bac2910ff04e5714631a541d92aec/ethcore/engines/authority-round/src/randomness.rs#L147).